README.md
# Zwave To MQTT

![GitHub package.json version](https://img.shields.io/github/package-json/v/OpenZWave/Zwave2Mqtt)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg?style=flat-square)](http://makeapullrequest.com)
[![MadeWithVueJs.com shield](https://madewithvuejs.com/storage/repo-shields/1897-shield.svg)](https://madewithvuejs.com/p/zwave2mqtt/shield-link)
[![MIT Licence](https://badges.frapsoft.com/os/mit/mit.png?v=103)](https://opensource.org/licenses/mit-license.php)
[![ci](https://github.com/OpenZWave/Zwave2Mqtt/workflows/ci/badge.svg?branch=master)](https://github.com/OpenZWave/Zwave2Mqtt/actions?query=workflow%3Aci+branch%3Amaster)
[![Docker Build](https://github.com/OpenZWave/Zwave2Mqtt/workflows/Docker%20Build/badge.svg?branch=master)](https://github.com/OpenZWave/Zwave2Mqtt/actions?query=workflow%3A%22Docker+Build%22+branch%3Amaster)
[![GitHub All Releases](https://img.shields.io/github/downloads/OpenZWave/Zwave2Mqtt/total)](https://github.com/OpenZWave/Zwave2Mqtt/releases)
[![Coverage Status](https://coveralls.io/repos/github/OpenZWave/Zwave2Mqtt/badge.svg?branch=master)](https://coveralls.io/github/OpenZWave/Zwave2Mqtt?branch=master)
[![Known Vulnerabilities](https://snyk.io/test/github/OpenZWave/Zwave2Mqtt/badge.svg?targetFile=package.json)](https://snyk.io/test/github/OpenZWave/Zwave2Mqtt?targetFile=package.json)
[![Dependencies Status](https://david-dm.org/OpenZWave/Zwave2Mqtt/status.svg)](https://david-dm.org/OpenZWave/Zwave2Mqtt)
[![devDependencies Status](https://david-dm.org/OpenZWave/Zwave2Mqtt/dev-status.svg)](https://david-dm.org/OpenZWave/Zwave2Mqtt?type=dev)
[![Total alerts](https://img.shields.io/lgtm/alerts/g/OpenZWave/Zwave2Mqtt.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/OpenZWave/Zwave2Mqtt/alerts/)
[![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/OpenZWave/Zwave2Mqtt.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/OpenZWave/Zwave2Mqtt/context:javascript)

[![Join channel](https://img.shields.io/badge/SLACK-zwave2mqtt.slack.com-red.svg?style=popout&logo=slack&logoColor=red)](https://join.slack.com/t/zwave2mqtt/shared_invite/enQtNjc4NjgyNjc3NDI2LTc3OGQzYmJlZDIzZTJhMzUzZWQ3M2Q3NThmMjY5MGY1MTc4NjFiOWZhZWE5YjNmNGE0OWRjZjJiMjliZGQyYmU 'Join channel')

[![Buy Me A Coffee](https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png)](https://www.buymeacoffee.com/MVg9wc2HE 'Buy Me A Coffee')

[![dockeri.co](https://dockeri.co/image/robertslando/zwave2mqtt)](https://hub.docker.com/r/robertslando/zwave2mqtt)

![OpenZWave](docs/OZW_Logo.png)
**TO**
![MQTT](docs/MQTT-Logo.png)

Fully configurable Zwave to MQTT **Gateway** and **Control Panel**.

- **Backend**: [NodeJS](https://nodejs.org/en/), [Express](https://expressjs.com/), [socket.io](https://github.com/socketio/socket.io), [Mqttjs](https://github.com/mqttjs/MQTT.js), [openzwave-shared](https://github.com/OpenZWave/node-openzwave-shared), [Webpack](https://webpack.js.org/)
- **Frontend**: [Vue](https://vuejs.org/), [socket.io](https://github.com/socketio/socket.io), [Vuetify](https://github.com/vuetifyjs/vuetify)

## !! ATTENTION

After a [discussion](https://github.com/OpenZWave/Zwave2Mqtt/issues/201) with Openzwave maintainer all issues related to **OZW 1.4** will be ignored and **automatically closed** as it isn't supported anymore, **please use OZW 1.6+**

## 📖 Table of contents

- [Zwave To MQTT](#zwave-to-mqtt)
  - [!! ATTENTION](#-attention)
  - [📖 Table of contents](#-table-of-contents)
  - [:electric_plug: Installation](#electric_plug-installation)
    - [DOCKER :tada: way](#docker-tada-way)
    - [Kubernetes way](#kubernetes-way)
    - [NodeJS or PKG version](#nodejs-or-pkg-version)
    - [Reverse Proxy Setup](#reverse-proxy-setup)
  - [:nerd_face: Development](#nerd_face-development)
    - [Developing against a different backend](#developing-against-a-different-backend)
  - [:wrench: Usage](#wrench-usage)
    - [Zwave](#zwave)
    - [MQTT](#mqtt)
    - [Gateway](#gateway)
      - [Special topics](#special-topics)
      - [Gateway values table](#gateway-values-table)
  - [:file_folder: Nodes Management](#file_folder-nodes-management)
    - [Add a node](#add-a-node)
    - [Remove a node](#remove-a-node)
    - [Replace failed node](#replace-failed-node)
    - [Remove a failed node](#remove-a-failed-node)
  - [:star: Features](#star-features)
  - [:robot: Home Assistant integration (BETA)](#robot-home-assistant-integration-beta)
    - [Components management](#components-management)
      - [Rediscover Node](#rediscover-node)
      - [Edit existing component](#edit-existing-component)
      - [Add new component](#add-new-component)
    - [Custom Components](#custom-components)
      - [Identify the Device id](#identify-the-device-id)
      - [Thermostats](#thermostats)
      - [Fans](#fans)
      - [Thermostats with Fans](#thermostats-with-fans)
  - [:gift: MQTT APIs](#gift-mqtt-apis)
    - [Zwave Events](#zwave-events)
      - [Example](#example)
    - [Zwave APIs](#zwave-apis)
      - [Custom APIs](#custom-apis)
    - [Set values](#set-values)
    - [Broadcast](#broadcast)
  - [:camera: Screenshots](#camera-screenshots)
    - [Settings](#settings)
    - [Control Panel](#control-panel)
    - [Groups associations](#groups-associations)
    - [Scenes](#scenes)
    - [Mesh](#mesh)
    - [Debug](#debug)
  - [Health check endpoints](#health-check-endpoints)
  - [Environment variables](#environment-variables)
  - [:question: FAQ](#question-faq)
  - [:pray: Thanks](#pray-thanks)
  - [:pencil: TODOs](#pencil-todos)
  - [:bowtie: Author](#bowtie-author)

## :electric_plug: Installation

### DOCKER :tada: way

```bash
# Using volumes as persistence
docker run --rm -it -p 8091:8091 --device=/dev/ttyACM0 --mount source=zwave2mqtt,target=/usr/src/app/store robertslando/zwave2mqtt:latest

# Using local folder as persistence
mkdir store
docker run --rm -it -p 8091:8091 --device=/dev/ttyACM0 -v $(pwd)/store:/usr/src/app/store robertslando/zwave2mqtt:latest

# As a service
wget https://raw.githubusercontent.com/openzwave/zwave2mqtt/master/docker/docker-compose.yml
docker-compose up
```

> Replace `/dev/ttyACM0` with your serial device

For more info about docker check [here](docker/README.md)

### Kubernetes way

```bash
kubectl apply -k https://raw.githubusercontent.com/OpenZWave/Zwave2Mqtt/master/kustomization.yaml
```

> You will almost certainly need to instead use this as a base, and then layer on top patches or resource customizations to your needs or just copy all the resources from the [kubernetes resources](./kubernetes) directory of this repo

### NodeJS or PKG version

1. Firstly you need to install the [Open-Zwave](https://github.com/OpenZWave/open-zwave) library on your system.

   ```sh
   cd ~
   git clone https://github.com/OpenZWave/open-zwave.git
   cd open-zwave && make && sudo make install
   sudo ldconfig
   export LD_LIBRARY_PATH=/usr/local/lib64
   sudo sed -i '$a LD_LIBRARY_PATH=/usr/local/lib64' /etc/environment
   ```

   For Raspberry check [here](https://github.com/OpenZWave/node-openzwave-shared/blob/master/README-raspbian.md#2-install-the-open-zwave-library-on-your-raspberry)

2. Test the library: go to openzwave directory `cd openzwave-*` and run the command

   `MinOZW /dev/ttyACM0`

   > replace `/dev/ttyACM0` with the USB port where your controller is connected

3. Now you can use the packaged version (you don't need NodeJS/npm installed) or clone this repo and build the project:

   - For the packaged version:

     ```sh
     cd ~
     mkdir Zwave2Mqtt
     cd Zwave2Mqtt
     # download latest version
     curl -s https://api.github.com/repos/OpenZWave/Zwave2Mqtt/releases/latest  \
     | grep "browser_download_url.*zip" \
     | cut -d : -f 2,3 \
     | tr -d \" \
     | wget -i -
     unzip zwave2mqtt-v*.zip
     ./zwave2mqtt
     ```

   - If you want to compile last code from github:

     ```sh
     git clone https://github.com/OpenZWave/Zwave2Mqtt
     cd Zwave2Mqtt
     npm install
     npm run build
     npm start
     ```

4. Open the browser <http://localhost:8091>

### Reverse Proxy Setup

If you need to setup ZWave To MQTT behind a reverse proxy that needs a _subpath_ to
work, take a look at [the reverse proxy configuraiton docs](docs/subpath.md).

## :nerd_face: Development

Developers who wants to debug the application have to open 2 terminals.

In first terminal run `npm run dev` to start webpack-dev for front-end developing and hot reloading at <http://localhost:8092>
(**THE PORT FOR DEVELOPING IS 8092**)

In the second terminal run `npm run dev:server` to start the backend server with inspect and auto restart features (if you don't have nodemon installed: `npm install -g nodemon`)

To package the application run `npm run pkg` command and follow the steps

### Developing against a different backend

By default running `npm run dev:server` will proxy the reequests to a backend listening on _localhost_ on port _8091_.

If you want to run the development frontend against a different backend you have the following environment variables
that you can use to redirect to a different backend:

- **SERVER_HOST**: [Default: 'localhost'] the hostname or IP of the backend server you want to use;
- **SERVER_PORT**: [Default: '8091'] the port of the backend server you want to use;
- **SERVER_SSL**: [Default: undefined] if set to a value it will use _https_/_wss_ to connect to the backend;
- **SERVER_URL**: [Default: use the other variables] the full URL for the backend API, IE: `https://zwavetomqtt.home.net:8443/`
- **SERVER_WS_URL**: [Default: use the other variables] the full URL for the backend Socket, IE: `wss://zwavetomqtt.home.net:8443/`

## :wrench: Usage

Firstly you need to open the browser at the link <http://localhost:8091> and edit the settings for Zwave, MQTT and the Gateway.

### Zwave

Zwave settings:


CHANGELOG.md
### Changelog

All notable changes to this project will be documented in this file. Dates are displayed in UTC.

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

#### [v4.1.1](https://github.com/OpenZWave/Zwave2Mqtt/compare/v4.1.0...v4.1.1)

> 2 December 2020

- chore(docker): downgrade to ozw 1.6.1520 #843 [`4258802`](https://github.com/OpenZWave/Zwave2Mqtt/commit/42588022f01f06efacbf73f9f1cec16655db972c)

#### [v4.1.0](https://github.com/OpenZWave/Zwave2Mqtt/compare/v4.0.6...v4.1.0)

> 30 November 2020

- docs: fix Table of Contents Links [`#842`](https://github.com/OpenZWave/Zwave2Mqtt/pull/842)
- feat: persistent settings for node table filter, sorting, show hidden nodes [`#831`](https://github.com/OpenZWave/Zwave2Mqtt/pull/831)
- fix: incorrect drying HVAC mode to dry [`#833`](https://github.com/OpenZWave/Zwave2Mqtt/pull/833)
- fix: incorrect drying HVAC mode to dry [`#833`](https://github.com/OpenZWave/Zwave2Mqtt/pull/833)
- fix: upgrade @babel/polyfill from 7.11.5 to 7.12.1 [`#818`](https://github.com/OpenZWave/Zwave2Mqtt/pull/818)
- fix: upgrade serialport from 9.0.1 to 9.0.2 [`#819`](https://github.com/OpenZWave/Zwave2Mqtt/pull/819)
- feat: node table filter [`#823`](https://github.com/OpenZWave/Zwave2Mqtt/pull/823)
- chore: bump chrisns/openzwave to 1.6.1545 [`#825`](https://github.com/OpenZWave/Zwave2Mqtt/pull/825)
- feat: persistent settings for node table filter, sorting, show hidden nodes (#831) [`#490`](https://github.com/OpenZWave/Zwave2Mqtt/issues/490)
- feat: Update last active on node messages [`#490`](https://github.com/OpenZWave/Zwave2Mqtt/issues/490)
- refactor: Introduce reusable settings compoment [`c101f0e`](https://github.com/OpenZWave/Zwave2Mqtt/commit/c101f0ec90328635496c93bb58e68a1a389242da)
- chore(deps): bump debug from 4.1.1 to 4.2.0 [`932a98a`](https://github.com/OpenZWave/Zwave2Mqtt/commit/932a98ab0a9aedfa2e21c096b58b9b13996f36db)
- refactor: Support different setting data types [`2e45e02`](https://github.com/OpenZWave/Zwave2Mqtt/commit/2e45e0290147c421d4ee67eebaaee54aaf067e63)
- feat: Store filters + sorting in local storage [`bda7208`](https://github.com/OpenZWave/Zwave2Mqtt/commit/bda720809f3b34863c0babf71087df0b9227e94b)
- chore(deps-dev): bump html-webpack-plugin from 4.4.1 to 4.5.0 [`5f9e07a`](https://github.com/OpenZWave/Zwave2Mqtt/commit/5f9e07af8ae7f9322794200191f26377f2f26699)
- Release 4.1.0 [`846cbde`](https://github.com/OpenZWave/Zwave2Mqtt/commit/846cbde113fd75dc911213c167ea5653693ca6a2)
- test: Fix test description wording [`8a96dce`](https://github.com/OpenZWave/Zwave2Mqtt/commit/8a96dceb29e9d3965d2ab9817888a8001c25750e)
- docs: Add docs to Settings module [`7717e46`](https://github.com/OpenZWave/Zwave2Mqtt/commit/7717e46f5a4fafce9ae3e2e255a6a069259c19d8)
- fix: prevent ignore discovery to dup values #837 [`e9553eb`](https://github.com/OpenZWave/Zwave2Mqtt/commit/e9553ebcea38a823edfa1edcf19d0f847355c0f8)
- fix: File locations and module style [`5ab71e5`](https://github.com/OpenZWave/Zwave2Mqtt/commit/5ab71e5a837662e3742ff89999c226d3a144992b)
- chore(deps-dev): bump ora from 5.0.0 to 5.1.0 [`a44e6a1`](https://github.com/OpenZWave/Zwave2Mqtt/commit/a44e6a1d7da3d6930ec891afab38e8bc7eef7a20)
- fix: Remove triggerStatusUpdate [`a378264`](https://github.com/OpenZWave/Zwave2Mqtt/commit/a378264c3e2cb0051da2adf593a8ec8106a551e3)
- fix: eslint errors [`638e292`](https://github.com/OpenZWave/Zwave2Mqtt/commit/638e29228173c0d4421946bb7ea6e3176ebc9448)
- fix: Check failures [`105d62a`](https://github.com/OpenZWave/Zwave2Mqtt/commit/105d62ae7f4f029f0beac69b7f9de1d788a66c8d)
- refactor: Simplify changes [`4bb2fe3`](https://github.com/OpenZWave/Zwave2Mqtt/commit/4bb2fe3bcfde49cf85b562bf10145ec36cdda9d5)
- fix: Use const for filtersStr [`7fccb37`](https://github.com/OpenZWave/Zwave2Mqtt/commit/7fccb37bcebdcf5951e576646c11ea167260f917)

#### [v4.0.6](https://github.com/OpenZWave/Zwave2Mqtt/compare/v4.0.5...v4.0.6)

> 4 November 2020

- feat(ui): save items per page in local storage [`#816`](https://github.com/OpenZWave/Zwave2Mqtt/pull/816)
- chore(deps): bump chrisns/openzwave in /docker [`#815`](https://github.com/OpenZWave/Zwave2Mqtt/pull/815)
- fix(hass): misuse drying state for Full Power mode [`#775`](https://github.com/OpenZWave/Zwave2Mqtt/pull/775)
- chore(deps): bump actions/setup-node from v1 to v2.1.2 [`#768`](https://github.com/OpenZWave/Zwave2Mqtt/pull/768)
- chore(deps): bump prismjs from 1.21.0 to 1.22.0 [`#778`](https://github.com/OpenZWave/Zwave2Mqtt/pull/778)
- chore(deps): bump socket.io-client from 2.3.0 to 2.3.1 [`#759`](https://github.com/OpenZWave/Zwave2Mqtt/pull/759)
- chore(deps): bump pascalgn/automerge-action from v0.11.0 to v0.12.0 [`#772`](https://github.com/OpenZWave/Zwave2Mqtt/pull/772)
- fix(ui): vuetify deprecation `.native` [`#797`](https://github.com/OpenZWave/Zwave2Mqtt/pull/797)

lib/Constants.js
module.exports = {
  // https://github.com/OpenZWave/open-zwave/blob/0d94c9427bbd19e47457578bccc60b16c6679b49/config/Localization.xml#L606
  _productionMap: {
    0: 'instant',
    1: 'total',
    2: 'today',
    3: 'time'
  },
  productionType (index) {
    return {
      sensor: 'energy_production',
      objectId: this._productionMap[index] || 'unknown',
      props: {
        device_class: index === 3 ? 'timestamp' : 'power'
      }
    }
  },
  // https://github.com/OpenZWave/open-zwave/blob/0d94c9427bbd19e47457578bccc60b16c6679b49/config/Localization.xml#L679
  // https://github.com/OpenZWave/open-zwave/blob/0d94c9427bbd19e47457578bccc60b16c6679b49/cpp/src/command_classes/Meter.cpp#L74
  _metersMap: {
    0: 'kwh', // electricity
    1: 'kvah',
    2: 'w',
    3: 'pulses',
    4: 'v',
    5: 'a',
    6: 'pf',
    8: 'kvar',
    9: 'kvarh',
    16: 'm3', // gas
    17: 'ft3',
    19: 'pulses',
    32: 'm3', // water
    33: 'ft3',
    34: 'gal',
    35: 'pulses',
    48: 'kwh', // heating (electricity)
    64: 'kwh' // cooling (electricity)
  },
  meterType (index) {
    let cfg = null
    if (index >= 16 && index < 32) {
      // gas
      cfg = this.sensorType(55)
    } else if (index >= 32 && index < 48) {
      // water
      cfg = this.sensorType(12)
    } else {
      // electricity
      cfg = this.sensorType(4)
    }

    cfg.objectId = (this._metersMap[index] || 'unknown') + '_meter'
    return cfg
  },
  // https://github.com/OpenZWave/open-zwave/blob/4478eea26b0e1a29184df0515a8034757258ff88/cpp/src/ValueIDIndexesDefines.def#L1068
  // https://github.com/OpenZWave/open-zwave/blob/05a096f75dddd27e3f8dc6af2afdb3cad3b4ebaa/config/Localization.xml#L7
  _alarmMap: {
    1: 'smoke',
    2: 'carbon_monoxide',
    3: 'carbon_dioxide',
    4: 'heat',
    5: 'water',
    6: 'access_control',
    7: 'home_security',
    8: 'power_management',
    9: 'system',
    10: 'emergency',
    11: 'clock',
    12: 'appliance',
    13: 'home_health',
    14: 'siren',
    15: 'water_valve',
    16: 'weather',
    17: 'irrigation',
    18: 'gas',
    256: 'previous_event_cleared',
    257: 'alert_location',
    512: 'type',
    513: 'level'
  },
  alarmType (index) {
    return this._alarmMap[index] || 'unknown_' + index
  },
  // https://github.com/OpenZWave/open-zwave/blob/0d94c9427bbd19e47457578bccc60b16c6679b49/config/SensorMultiLevelCCTypes.xml
  _sensorMap: {
    temperature: {
      1: 'air',
      11: 'dew_point',
      23: 'water',
      24: 'soil',
      34: 'target',
      62: 'boiler_water',
      63: 'domestic_hot_water',
      64: 'outside',
      65: 'exhaust',
      72: 'return_air',
      73: 'supply_air',
      74: 'condenser_coil',
      75: 'evaporator_coil',
      76: 'liquid_line',
      77: 'discharge_line',
      80: 'defrost',
      props: {
        device_class: 'temperature'
      }
    },
    illuminance: {
      3: '', // illuminance
      props: {
        device_class: 'illuminance'
      }
    },
    electricity: {
      4: 'power',
      15: 'voltage',
      16: 'current',
      28: 'resistivity',
      29: 'conductivity',
      props: {
        device_class: 'power'
      }
    },
    humidity: {
      5: 'air', // humidity
      41: 'soil',
      props: {
        device_class: 'humidity'
      }
    },
    speed: {
      6: 'velocity', // ?
      52: 'x_acceleration',
      53: 'y_acceleration',
      54: 'z_acceleration',
      props: {
        icon: 'mdi:speedometer'
      }
    },
    direction: {
      7: '', // direction
      21: 'angle_position', // ?
      props: {
        icon: 'mdi:compass'
      }
    },
    pressure: {
      8: 'atmospheric',
      9: 'barometric',
      57: 'water',
      props: {
        device_class: 'pressure'
      }
    },
    sun: {
      10: 'solar_radiation',
      27: 'ultraviolet',
      props: {
        icon: 'mdi:white-balance-sunny'
      }
    },
    water: {
      12: 'rain_rate',
      13: 'tide_level',
      19: 'tank_capacity',
      56: 'flow',
      props: {
        icon: 'mdi:water'
      }
    },
    weight: {
      14: '', // weight
      46: 'muscle_mass',
      47: 'bone_mass',
      48: 'fat_mass',
      49: 'total_body_water',
      51: 'body_mass_index',
      66: 'water_chlorine',
      67: 'water_acidity',
      68: 'water_oxidation_potential',
      props: {
        icon: 'mdi:weight'
      }
    },
    gas: {
      17: 'carbon_dioxide',
      40: 'carbon_monoxide',
      55: 'smoke_density',
      props: {
        icon: 'mdi:thought-bubble'
      }
    },
    air: {
      18: 'flow',
      props: {
        icon: 'mdi:air-filter'
      }
    },
    frequency: {
      22: 'rotation',
      32: '',
      props: {}
    },
    sound: {
      30: 'loudness',
      props: {
        icon: 'mdi:volume-high'
      }
    },
    signal: {
      58: 'strength',
      props: {
        device_class: 'signal_strength'
      }
    },
    timestamp: {
      33: '', // time,
      props: {
        device_class: 'timestamp'
      }
    },
    heart: {
      44: 'rate',
      45: 'blood_pressure',
      50: 'basic_metabolic_rate',
      60: 'respiratory_rate',
      69: 'lf_lh_ratio',
      props: {
        icon: 'mdi:heart'
      }
    },
    generic: {
      2: 'general_purpose',
      20: 'distance',
      25: 'seismic_intensity',
      26: 'seismic_magnitude',
      31: 'moisture',
      35: 'particulate_matter_25',
      36: 'formaldehyde',
      37: 'radon_concentration',
      38: 'methane_density',
      39: 'volatile_organic_compound',
      42: 'soil_reactivity',
      43: 'soil_salinity',
      59: 'particulate_matter',
      61: 'relative_modulation',
      70: 'motion_direction',
      71: 'applied_force',
      78: 'suction',
      79: 'discharge',
      81: 'ozone',
      82: 'sulfur_dioxide',
      83: 'nitrogen_dioxide',
      84: 'ammonia',
      85: 'lead',
      86: 'particulate_matter'
    }
  },
  sensorType (index) {
    const sensorType = {
      sensor: 'generic',
      objectId: 'unknown_' + index,
      props: {}
    }

    for (const sensor in this._sensorMap) {
      const objectId = this._sensorMap[sensor][index]
      if (objectId !== undefined) {
        sensorType.sensor = sensor
        sensorType.objectId = objectId
        sensorType.props = this._sensorMap[sensor].props || {}
        break
      }
    }

    return sensorType
  },
  _commandClassMap: {
    0x00: 'no_operation',
    0x20: 'basic',
    0x21: 'controller_replication',
    0x22: 'application_status',
    0x23: 'zip_services',
    0x24: 'zip_server',
    0x25: 'switch_binary',
    0x26: 'switch_multilevel',
    0x27: 'switch_all',
    0x28: 'switch_toggle_binary',
    0x29: 'switch_toggle_multilevel',
    0x2a: 'chimney_fan',
    0x2b: 'scene_activation',
    0x2c: 'scene_actuator_conf',
    0x2d: 'scene_controller_conf',
    0x2e: 'zip_client',
    0x2f: 'zip_adv_services',
    0x30: 'sensor_binary',
    0x31: 'sensor_multilevel',
    0x32: 'meter',
    0x33: 'color',
    0x34: 'zip_adv_client',
    0x35: 'meter_pulse',
    0x3c: 'meter_tbl_config',
    0x3d: 'meter_tbl_monitor',
    0x3e: 'meter_tbl_pulse',
    0x38: 'thermostat_heating',
    0x40: 'thermostat_mode',
    0x42: 'thermostat_operating_state',
    0x43: 'thermostat_setpoint',
    0x44: 'thermostat_fan_mode',
    0x45: 'thermostat_fan_state',
    0x46: 'climate_control_schedule',
    0x47: 'thermostat_setback',
    0x4c: 'door_lock_logging',
    0x4e: 'schedule_entry_lock',
    0x50: 'basic_window_covering',
    0x51: 'mtp_window_covering',
    0x56: 'crc16_encap',
    0x5a: 'device_reset_locally',
    0x5b: 'central_scene',
    0x5e: 'zwave_plus_info',
    0x5d: 'antitheft',
    0x60: 'multi_instance',
    0x62: 'door_lock',
    0x63: 'user_code',
    0x66: 'barrier_operator',
    0x70: 'configuration',
    0x71: 'alarm',
    0x72: 'manufacturer_specific',
    0x73: 'powerlevel',
    0x75: 'protection',
    0x76: 'lock',
    0x77: 'node_naming',
    0x79: 'sound_switch',
    0x7a: 'firmware_update_md',
    0x7b: 'grouping_name',
    0x7c: 'remote_association_activate',
    0x7d: 'remote_association',
    0x80: 'battery',
    0x81: 'clock',
    0x82: 'hail',
    0x84: 'wake_up',
    0x85: 'association',
    0x86: 'version',
    0x87: 'indicator',
    0x88: 'proprietary',
    0x89: 'language',
    0x8a: 'time',
    0x8b: 'time_parameters',
    0x8c: 'geographic_location',
    0x8d: 'composite',
    0x8e: 'multi_instance_association',
    0x8f: 'multi_cmd',
    0x90: 'energy_production',
    0x91: 'manufacturer_proprietary',
    0x92: 'screen_md',
    0x93: 'screen_attributes',
    0x94: 'simple_av_control',
    0x95: 'av_content_directory_md',
    0x96: 'av_renderer_status',
    0x97: 'av_content_search_md',
    0x98: 'security',
    0x99: 'av_tagging_md',
    0x9a: 'ip_configuration',
    0x9b: 'association_command_configuration',
    0x9c: 'sensor_alarm',
    0x9d: 'silence_alarm',
    0x9e: 'sensor_configuration',
    0xef: 'mark',
    0xf0: 'non_interoperable'
  },
  commandClass (cmd) {
    return this._commandClassMap[cmd] || 'unknownClass_' + cmd
  },
  _genericDeviceClassMap: {
    // https://github.com/OpenZWave/open-zwave/blob/master/config/device_classes.xml
    // https://github.com/home-assistant/core/blob/dev/homeassistant/components/zwave/const.py#L196
    // 0x00: specific_type_not_used // Available in all Generic types
    0x03: {
      generic: 'generic_type_av_control_point',
      specific: {
        0x12: 'specific_type_doorbell',
        0x04: 'specific_type_satellite_receiver',
        0x11: 'specific_type_satellite_receiver_v2'
      }
    },
    0x04: {
      generic: 'generic_type_display',
      specific: {
        0x01: 'specific_type_simple_display'
      }
    },
    0x40: {
      generic: 'generic_type_entry_control',
      specific: {
        0x01: 'specific_type_door_lock',
        0x02: 'specific_type_advanced_door_lock',
        0x03: 'specific_type_secure_keypad_door_lock',
        0x04: 'specific_type_secure_keypad_door_lock_deadbolt',
        0x05: 'specific_type_secure_door',
        0x06: 'specific_type_secure_gate',
        0x07: 'specific_type_secure_barrier_addon',
        0x08: 'specific_type_secure_barrier_open_only',
        0x09: 'specific_type_secure_barrier_close_only',
        0x10: 'specific_type_secure_lockbox',
        0x11: 'specific_type_secure_keypad'
      }
    },
    0x01: {
      generic: 'generic_type_generic_controller',
      specific: {
        0x01: 'specific_type_portable_controller',
        0x02: 'specific_type_portable_scene_controller',
        0x03: 'specific_type_portable_installer_tool',
        0x04: 'specific_type_control_av',
        0x06: 'specific_type_control_simple'
      }
    },
    0x31: {
      generic: 'generic_type_meter',
      specific: {
        0x01: 'specific_type_simple_meter',
        0x02: 'specific_type_adv_energy_control',
        0x03: 'specific_type_whole_home_meter_simple'
      }
    },
    0x30: {
      generic: 'generic_type_meter_pulse',
      specific: {}
    },
    0xff: {
      generic: 'generic_type_non_interoperable',
      specific: {}
    },
    0x0f: {
      generic: 'generic_type_repeater_slave',
      specific: {
        0x01: 'specific_type_repeater_slave',
        0x02: 'specific_type_virtual_node'
      }
    },
    0x17: {
      generic: 'generic_type_security_panel',
      specific: {
        0x01: 'specific_type_zoned_security_panel'
      }
    },
    0x50: {
      generic: 'generic_type_semi_interoperable',
      specific: {
        0x01: 'specific_type_energy_production'
      }
    },
    0xa1: {
      generic: 'generic_type_sensor_alarm',
      specific: {
        0x05: 'specific_type_adv_zensor_net_alarm_sensor',
        0x10: 'specific_type_adv_zensor_net_smoke_sensor',
        0x01: 'specific_type_basic_routing_alarm_sensor',
        0x06: 'specific_type_basic_routing_smoke_sensor',
        0x03: 'specific_type_basic_zensor_net_alarm_sensor',
        0x08: 'specific_type_basic_zensor_net_smoke_sensor',
        0x02: 'specific_type_routing_alarm_sensor',
        0x07: 'specific_type_routing_smoke_sensor',
        0x04: 'specific_type_zensor_net_alarm_sensor',
        0x09: 'specific_type_zensor_net_smoke_sensor',
        0x11: 'specific_type_alarm_sensor'
      }
    },
    0x20: {
      generic: 'generic_type_sensor_binary',
      specific: {
        0x01: 'specific_type_routing_sensor_binary'
      }
    },
    0x21: {
      generic: 'generic_type_sensor_multilevel',
      specific: {
        0x01: 'specific_type_routing_sensor_multilevel',
        0x02: 'specific_type_chimney_fan'
      }
    },
    0x02: {
      generic: 'generic_type_static_controller',
      specific: {
        0x01: 'specific_type_pc_controller',
        0x02: 'specific_type_scene_controller',
        0x03: 'specific_type_static_installer_tool',
        0x04: 'specific_type_set_top_box',
        0x05: 'specific_type_sub_system_controller',
        0x06: 'specific_type_tv',
        0x07: 'specific_type_gateway'
      }
    },
    0x10: {
      generic: 'generic_type_switch_binary',
      specific: {
        0x01: 'specific_type_power_switch_binary',
        0x03: 'specific_type_scene_switch_binary',
        0x04: 'specific_type_power_strip',
        0x05: 'specific_type_siren',
        0x06: 'specific_type_valve_open_close',
        0x02: 'specific_type_color_tunable_binary',
        0x07: 'specific_type_irrigation_controller'
      }
    },
    0x11: {
      generic: 'generic_type_switch_multilevel',
      specific: {
        0x05: 'specific_type_class_a_motor_control',
        0x06: 'specific_type_class_b_motor_control',
        0x07: 'specific_type_class_c_motor_control',
        0x03: 'specific_type_motor_multiposition',
        0x01: 'specific_type_power_switch_multilevel',
        0x04: 'specific_type_scene_switch_multilevel',
        0x08: 'specific_type_fan_switch',
        0x02: 'specific_type_color_tunable_multilevel'
      }
    },
    0x12: {
      generic: 'generic_type_switch_remote',
      specific: {
        0x01: 'specific_type_remote_binary',
        0x02: 'specific_type_remote_multilevel',
        0x03: 'specific_type_remove_toggle_binary',
        0x04: 'specific_type_remote_toggle_multilevel'
      }
    },
    0x13: {
      generic: 'generic_type_switch_toggle',
      specific: {
        0x01: 'specific_type_switch_toggle_binary',
        0x02: 'specific_type_switch_toggle_multilevel'
      }
    },
    0x08: {
      generic: 'generic_type_thermostat',
      specific: {
        0x03: 'specific_type_setback_schedule_thermostat',
        0x05: 'specific_type_setback_thermostat',
        0x04: 'specific_type_setpoint_thermostat',
        0x02: 'specific_type_thermostat_general',
        0x06: 'specific_type_thermostat_general_v2',
        0x01: 'specific_type_thermostat_heating'
      }
    },
    0x16: {
      generic: 'generic_type_ventilation',
      specific: {
        0x01: 'specific_type_residential_hrv'
      }
    },
    0x09: {
      generic: 'generic_type_windows_covering',
      specific: {
        0x01: 'specific_type_simple_window_covering'
      }
    },
    0x15: {
      generic: 'generic_type_zip_node',
      specific: {
        0x02: 'specific_type_zip_adv_node',
        0x01: 'specific_type_zip_tun_node'
      }
    },
    0x18: {
      generic: 'generic_type_wall_controller',
      specific: {
        0x01: 'specific_type_basic_wall_controller'
      }
    },
    0x05: {
      generic: 'generic_type_network_extender',
      specific: {
        0x01: 'specific_type_secure_extender'
      }
    },
    0x06: {
      generic: 'generic_type_appliance',
      specific: {
        0x01: 'specific_type_general_appliance',
        0x02: 'specific_type_kitchen_appliance',
        0x03: 'specific_type_laundry_appliance'
      }
    },
    0x07: {
      generic: 'generic_type_sensor_notification',
      specific: {
        0x01: 'specific_type_notification_sensor'
      }
    }
  },
  genericDeviceClassAttributes (cls) {
    return this._genericDeviceClassMap[cls]
  },
  genericDeviceClass (cls) {
    var clsAttr = this.genericDeviceClassAttributes(cls)
    if (clsAttr) {
      return clsAttr.generic
    } else {
      return 'unknownGenericDeviceType_' + cls
    }
  },
  specificDeviceClass (genericCls, specificCls) {
    var clsAttr = this.genericDeviceClassAttributes(genericCls)
    if (clsAttr) {
      return (
        clsAttr.specific[specificCls] ||
        'unknownSpecificDeviceType_' + specificCls
      )
    } else {
      return 'unknownGenericDeviceType_' + genericCls
    }
  }
}
